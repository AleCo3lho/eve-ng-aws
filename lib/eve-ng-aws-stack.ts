import * as cdk from "aws-cdk-lib";
import {
  AutoScalingGroup,
  BlockDeviceVolume,
} from "aws-cdk-lib/aws-autoscaling";
import {
  InstanceClass,
  InstanceSize,
  InstanceType,
  MachineImage,
  Peer,
  Port,
  SecurityGroup,
  SubnetType,
  UserData,
  Vpc,
} from "aws-cdk-lib/aws-ec2";
import { Bucket } from "aws-cdk-lib/aws-s3";
import { Construct } from "constructs";

export class EveNgAwsStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const vpc = new Vpc(this, "EveNGVpc", {
      maxAzs: 1,
      subnetConfiguration: [
        {
          name: "PublicSubNet",
          subnetType: SubnetType.PUBLIC,
        },
      ],
      natGateways: 0,
    });
    vpc.applyRemovalPolicy(cdk.RemovalPolicy.DESTROY);

    const s3 = new Bucket(this, "bucketWithImages");

    const userData = UserData.forLinux();
    userData.addCommands(
      `sudo -i\n`,
      `set -o xtrace\n`,
      `resize2fs /dev/sda1\n`,
      `sed -i -e "s/.*PermitRootLogin .*/PermitRootLogin yes/" /etc/ssh/sshd_config\n`,
      `sed -i -e "s/.*PasswordAuthentication .*/PasswordAuthentication yes/" /etc/ssh/sshd_config\n`,
      `service ssh restart\n`,
      `wget -O - http://www.eve-ng.net/jammy/eczema@ecze.com.gpg.key | gpg --dearmor -o /etc/apt/trusted.gpg.d/eve-ng.gpg\n`,
      `echo "deb [arch=amd64] http://www.eve-ng.net/jammy jammy main" > /etc/apt/sources.list.d/eve-ng.list\n`,
      `apt-get update\n`,
      `apt-get upgrade\n`,
      `apt-get -y install software-properties-common python3-pip\n`,
      `DEBIAN_FRONTEND=noninteractive apt-get -y install eve-ng\n`,
      `/opt/unetlab/wrappers/unl_wrapper -a fixpermissions\n`,
    );

    const securityGroup = new SecurityGroup(this, "EveNgSG", {
      vpc: vpc,
      allowAllOutbound: true,
    });

    securityGroup.addIngressRule(Peer.anyIpv4(), Port.allTraffic());
    securityGroup.applyRemovalPolicy(cdk.RemovalPolicy.DESTROY);

    const asg = new AutoScalingGroup(this, "EveNGASG", {
      vpc: vpc,
      ssmSessionPermissions: true,
      securityGroup: securityGroup,
      desiredCapacity: 1,
      userData: userData,
      machineImage: MachineImage.genericLinux({
        "us-east-1": "ami-0a0e5d9c7acc336f1",
      }),
      blockDevices: [
        {
          deviceName: "/dev/sda1",
          volume: BlockDeviceVolume.ebs(100),
        },
      ],
      instanceType: InstanceType.of(InstanceClass.C5, InstanceSize.METAL),
      associatePublicIpAddress: true,
      spotPrice: "1.50",
    });
    asg.applyRemovalPolicy(cdk.RemovalPolicy.DESTROY);
    s3.grantReadWrite(asg.role);
  }
}
