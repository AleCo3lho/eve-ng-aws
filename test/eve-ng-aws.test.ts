import * as cdk from "aws-cdk-lib";
import { EveNgAwsStack } from "../lib/eve-ng-aws-stack";
import { Template } from "aws-cdk-lib/assertions";

test("snapshot for EveNGStack matches previous state", () => {
  const app = new cdk.App();
  const stack = new EveNgAwsStack(app, "TestEveNGStack");

  const template = Template.fromStack(stack);

  expect(template.toJSON()).toMatchSnapshot();
});
