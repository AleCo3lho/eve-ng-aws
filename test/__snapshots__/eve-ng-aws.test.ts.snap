// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot for EveNGStack matches previous state 1`] = `
{
  "Mappings": {
    "EveNGASGAmiMapA8284209": {
      "us-east-1": {
        "ami": "ami-0a0e5d9c7acc336f1",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "EveNGASGB28FC224": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "DesiredCapacity": "1",
        "LaunchConfigurationName": {
          "Ref": "EveNGASGLaunchConfigCDF0573A",
        },
        "MaxSize": "1",
        "MinSize": "1",
        "Tags": [
          {
            "Key": "Name",
            "PropagateAtLaunch": true,
            "Value": "TestEveNGStack/EveNGASG",
          },
        ],
        "VPCZoneIdentifier": [
          {
            "Ref": "EveNGVpcPublicSubNetSubnet1SubnetFF592884",
          },
        ],
      },
      "Type": "AWS::AutoScaling::AutoScalingGroup",
      "UpdatePolicy": {
        "AutoScalingScheduledAction": {
          "IgnoreUnmodifiedGroupSizeProperties": true,
        },
      },
      "UpdateReplacePolicy": "Delete",
    },
    "EveNGASGInstanceProfileED15502C": {
      "Properties": {
        "Roles": [
          {
            "Ref": "EveNGASGInstanceRole88B26E6E",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "EveNGASGInstanceRole88B26E6E": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "ec2.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AmazonSSMManagedInstanceCore",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestEveNGStack/EveNGASG",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "EveNGASGLaunchConfigCDF0573A": {
      "DependsOn": [
        "EveNGASGInstanceRole88B26E6E",
      ],
      "Properties": {
        "AssociatePublicIpAddress": true,
        "BlockDeviceMappings": [
          {
            "DeviceName": "/dev/sda1",
            "Ebs": {
              "VolumeSize": 100,
            },
          },
        ],
        "IamInstanceProfile": {
          "Ref": "EveNGASGInstanceProfileED15502C",
        },
        "ImageId": {
          "Fn::FindInMap": [
            "EveNGASGAmiMapA8284209",
            {
              "Ref": "AWS::Region",
            },
            "ami",
          ],
        },
        "InstanceType": "c5.metal",
        "SecurityGroups": [
          {
            "Fn::GetAtt": [
              "EveNgSG8D618850",
              "GroupId",
            ],
          },
        ],
        "SpotPrice": "1.30",
        "UserData": {
          "Fn::Base64": "#!/bin/bash
sudo -i

set -o xtrace

resize2fs /dev/sda1

sed -i -e "s/.*PermitRootLogin .*/PermitRootLogin yes/" /etc/ssh/sshd_config

sed -i -e "s/.*PasswordAuthentication .*/PasswordAuthentication yes/" /etc/ssh/sshd_config

service ssh restart

wget -O - http://www.eve-ng.net/jammy/eczema@ecze.com.gpg.key | gpg --dearmor -o /etc/apt/trusted.gpg.d/eve-ng.gpg

echo "deb [arch=amd64] http://www.eve-ng.net/jammy jammy main" > /etc/apt/sources.list.d/eve-ng.list

apt-get update

apt-get upgrade

apt-get -y install software-properties-common python3-pip

DEBIAN_FRONTEND=noninteractive apt-get -y install eve-ng

pip install gdown

gdown https://drive.google.com/drive/folders/1ORWpJOO-8B-tsQe95gnhEqzQ5vZl2CMm -O /root/eveng/os --folder

mkdir -p /opt/unetlab/addons/dynamips/
mv /root/eveng/os/image/* /opt/unetlab/addons/dynamips/

mkdir -p /opt/unetlab/addons/qemu/
mv /root/eveng/os/qemu/* /opt/unetlab/addons/qemu/

mkdir -p /opt/unetlab/addons/iol/
mv /root/eveng/os/IOL/* /opt/unetlab/addons/iol/

cd /opt/unetlab/addons/iol/

cat *.tar.gz | tar zxvf - -i

/opt/unetlab/wrappers/unl_wrapper -a fixpermissions
",
        },
      },
      "Type": "AWS::AutoScaling::LaunchConfiguration",
    },
    "EveNGVpc1017EA53": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestEveNGStack/EveNGVpc",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
      "UpdateReplacePolicy": "Delete",
    },
    "EveNGVpcIGWB0E7859E": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestEveNGStack/EveNGVpc",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "EveNGVpcPublicSubNetSubnet1DefaultRoute2654F977": {
      "DependsOn": [
        "EveNGVpcVPCGWDA51C968",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "EveNGVpcIGWB0E7859E",
        },
        "RouteTableId": {
          "Ref": "EveNGVpcPublicSubNetSubnet1RouteTable6A240009",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "EveNGVpcPublicSubNetSubnet1RouteTable6A240009": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestEveNGStack/EveNGVpc/PublicSubNetSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "EveNGVpc1017EA53",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "EveNGVpcPublicSubNetSubnet1RouteTableAssociation2D6B4EE5": {
      "Properties": {
        "RouteTableId": {
          "Ref": "EveNGVpcPublicSubNetSubnet1RouteTable6A240009",
        },
        "SubnetId": {
          "Ref": "EveNGVpcPublicSubNetSubnet1SubnetFF592884",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "EveNGVpcPublicSubNetSubnet1SubnetFF592884": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.0.0/16",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "PublicSubNet",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "Name",
            "Value": "TestEveNGStack/EveNGVpc/PublicSubNetSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "EveNGVpc1017EA53",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "EveNGVpcVPCGWDA51C968": {
      "Properties": {
        "InternetGatewayId": {
          "Ref": "EveNGVpcIGWB0E7859E",
        },
        "VpcId": {
          "Ref": "EveNGVpc1017EA53",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
    "EveNgSG8D618850": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "GroupDescription": "TestEveNGStack/EveNgSG",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "SecurityGroupIngress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "from 0.0.0.0/0:ALL TRAFFIC",
            "IpProtocol": "-1",
          },
        ],
        "VpcId": {
          "Ref": "EveNGVpc1017EA53",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
